File: 078. Subsets.swift
class Solution {
    func subsets(_ nums: [Int]) -> [[Int]] {
        let sorted = nums.sorted()
        return [[]] + subTree(tempRoot: [], nums: sorted, startIndex: 0)
    }
    
    func subTree(tempRoot: [Int], nums: [Int], startIndex: Int) -> [[Int]] {
        
        var result = [[Int]]()
        
        for index in startIndex ..< nums.count {
            let subTempRoot = tempRoot + [nums[index]]
            result = result + [subTempRoot] + subTree(tempRoot: subTempRoot, nums: nums, startIndex: index+1)
        }
        
        return result
    }
}

File: 078.Subsets.java
/*
    Given a set of distinct integers, nums, return all possible subsets (the power set).

    Note: The solution set must not contain duplicate subsets.

    For example,
    If nums = [1,2,3], a solution is:

    [
      [3],
      [1],
      [2],
      [1,2,3],
      [1,3],
      [2,3],
      [1,2],
      []
    ]
     */

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        if (nums.length == 0)
            return result;
        result.add(new ArrayList<Integer>());

        for (int i=0; i<nums.length; i++) {
            int size = result.size();
            for (int j=0; j<size; j++) {
                List<Integer> temp = new ArrayList<>(result.get(j));
                temp.add(nums[i]);
                result.add(temp);
            }
        }

        return result;
    }
